src/NuGet.Core/NuGet.Versioning/NuGetVersionFactory.cs

```diff
@@ -112,6 +104,77 @@ int IndexOf(string str, char c)
             }
         }
 
+        private static bool TryGetNormalizedVersion(string str, [NotNullWhen(true)] out Version? version)
+        {
+            if (!string.IsNullOrEmpty(str) && GetNextSection(str, 0, out int endIndex, out int major))
+            {
+                int build = 0;
+                int revision = 0;
+
+                // check for all the possible parts of the version string. If endIndex is less than the end of
+                // the string, the input string was invalid (e.g. "1.2.3.4.5").
+                bool validString = GetNextSection(str, endIndex + 1, out endIndex, out int minor) &&
+                    GetNextSection(str, endIndex + 1, out endIndex, out build) &&
+                    GetNextSection(str, endIndex + 1, out endIndex, out revision) &&
+                    endIndex == str.Length;
+
+                if (validString)
+                {
+                    version = new Version(major, Math.Max(0, minor), Math.Max(0, build), Math.Max(0, revision));
```

Comment:
I'm unsure if this matters, but our current code calls `Version.TryParse()` which returns a `Version` with only the appropriate version components set.  So `"1.0"` would return a version with `-1` as the build and revision.  This code now always calls the `Version` constructor with values of 0.  

To keep the current behavior, it would only need to specify `0` for the minor if its not specified but pass `-1` for the build and revision.

Reply:
Could be wrong, but it looked like the call to `NormalizeVersionValue` ensures that build and revision are set to a minimum of zero. 


        internal static Version NormalizeVersionValue(Version version)
        {
            var normalized = version;

            if (version.Build < 0
                || version.Revision < 0)
            {
                normalized = new Version(
                    version.Major,
                    version.Minor,
                    Math.Max(version.Build, 0),
                    Math.Max(version.Revision, 0));
            }

            return normalized;
        }


`NormalizeVersionValue` allocates a new `Version` instance to do this that I was hoping to avoid.

Reply:
Ah sorry I didn't realize that!